pkg/aws/ec2.go

```diff
@@ -252,11 +261,84 @@ func (r AwsEc2DynamicConfig) SshUser() string {
 	return "ec2-user"
 }
 
-// TODO: implement this function.
+// GetState returns ec's instanceId VM state from AWS. See https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_InstanceState.html for
+// valid states.
 func (ec AwsEc2DynamicConfig) GetState(kubeClient client.Client, ctx context.Context, instanceId cloud.InstanceIdentifier) (string, error) {
 	return "ACTIVE", nil
 }
 
+// CleanUpVms deletes any VMs in the ec cloud that are associated with a non-existing Tekton TaskRun.
+// Each VM instance should have a tag with the associated TaskRun's namespace and name. This is compared to
+// existingTaskRuns, which is a map of namespaces to a list of TaskRuns in that namespace, to determine if this
+// VM's TaskRun still exists.
+func (ec AwsEc2DynamicConfig) CleanUpVms(kubeClient client.Client, ctx context.Context, existingTaskRuns map[string][]string) error {
+	log := logr.FromContextOrDiscard(ctx)
+	log.Info("Attempting to clean up orphaned AWS EC2 instances")
+
+	// Create an EC2 client
+	ec2Client, err := ec.createClient(kubeClient, ctx)
+	if err != nil {
+		return fmt.Errorf("failed to create an EC2 client: %w", err)
+	}
+
+	// Retrieve instances
+	instancesOutput, err := ec2Client.DescribeInstances(
+		ctx,
+		&ec2.DescribeInstancesInput{
+			Filters: []types.Filter{
+				{Name: aws.String("tag:" + MultiPlatformManaged), Values: []string{"true"}},
+			},
+		},
+	)
+	if err != nil {
+		log.Error(err, "failed to retrieve EC2 instances")
+		return fmt.Errorf("failed to retrieve EC2 instances: %w", err)
+	}
+
+	// Iterate over all VM instances
+	for _, reservation := range instancesOutput.Reservations {
+		for _, instance := range reservation.Instances {
+			// Iterate over this instance's tags
+			for _, instanceTag := range instance.Tags {
+				if *instanceTag.Key != cloud.TaskRunTagKey {
+					continue
+				}
+				err := cloud.ValidateTaskRunId(*instanceTag.Key)
+				if err != nil {
+					msg := fmt.Sprintf("invalid TaskRun ID; skipping instance %s...", *instance.InstanceId)
+					log.Error(err, msg)
+					break
+				}
+
+				// Try to find this instance's TaskRun
+				taskRunExists := false
+				taskRunNamespace := strings.Split((*instanceTag.Value), ":")[0]
+				taskRuns, ok := existingTaskRuns[taskRunNamespace]
+				if ok {
+					instanceTaskRunName := strings.Split((*instanceTag.Value), ":")[1]
+					for _, existingTaskRunName := range taskRuns {
+						if instanceTaskRunName == existingTaskRunName {
+							taskRunExists = true
+							break
+						}
+					}
+				}
+
+				// Delete the VM instance if the TaskRun namespace or TaskRun does not exist
+				if !taskRunExists {
+					err := ec.TerminateInstance(kubeClient, ctx, cloud.InstanceIdentifier(*instance.InstanceId))
+					if err != nil {
+						log.Error(err, "failed to terminate instance", "instanceId", *instance.InstanceId)
+						return fmt.Errorf("failed to terminate instance %s: %w", *instance.InstanceId, err)
+					}
+				}
```

Comment:
we can simplify this a bit

```suggestion
				if *instanceTag.Key != cloud.TaskRunTagKey {
					continue
				}

				if err := cloud.ValidateTaskRunId(*instanceTag.Key); err != nil {
					msg := fmt.Sprintf("invalid TaskRun ID; skipping instance %s...", *instance.InstanceId)
					log.Error(err, msg)
					break
				}

				// Try to find this instance's TaskRun
				its := strings.Split((*instanceTag.Value), ":")
				taskRunNamespace, instanceTaskRunName := its[0], its[1]
				taskRuns, ok := existingTaskRuns[taskRunNamespace]
				if !ok {
					continue
				}

				// Delete the VM instance if the TaskRun namespace or TaskRun does not exist
				if !slices.Contains(taskRuns, instanceTaskRunName) {
					if err := ec.TerminateInstance(kubeClient, ctx, cloud.InstanceIdentifier(*instance.InstanceId)); err != nil {
						log.Error(err, "failed to terminate instance", "instanceId", *instance.InstanceId)
						return fmt.Errorf("failed to terminate instance %s: %w", *instance.InstanceId, err)
					}
				}
```

Reply:
Much better, thanks! I implemented these changes.